### PR TITLE
Fix the execution method of the context

### PR DESCRIPTION
**Pull request by neithernut**
evecve & co. get a NULL-terminated list with the first "argument" being
the command itself (it's a convention). The original code had a few
issues (e.g. it failed to accept commands without arguments). This
pathch fixes those issues by nearly completely rewriting the function.

### CODE REVIEW

**Comment by neithernut** [comment]
Incidently, this allows you to start a `weston-terminal` via a transaction.

File: src/context.c [hunk 1]
```diff
@@ -189,37 +189,41 @@ func_exec(
...
     enum ws_value_type iter_type;
 
     // count number of arguments and do type checking meanwhile
-    for (n = 0; iter; n++) {
-        iter_type = ws_value_get_type(&stack[n].value);
-        if (iter_type != WS_VALUE_TYPE_STRING) {
-            return -EINVAL;
-        }
+    while ((iter_type = ws_value_get_type(&stack[n].value)) ==
+            WS_VALUE_TYPE_STRING) {
```
Comment (matthiasbeyer) [inline src/context.c:197]: Could you possibly use the `ITERARTE_ARGS_TYPE()` macro here?

File: src/context.c [hunk 2]
```diff
@@ -189,37 +189,41 @@ func_exec(
...
 
-        iter = iter_type != WS_VALUE_TYPE_NONE;
+    // check whether there is a command (argument 0)
+    if ((iter_type != WS_VALUE_TYPE_NONE) || !n) {
+        return -EINVAL;
     }
 
-    const char* args[n];
+    // allocate room for a NULL-terminated list of args
+    const char* args[n+1];
```
Comment (matthiasbeyer) [inline src/context.c]: Please add spaces around the `+` here.

File: src/context.c (hunk 1 above)
Comment (neithernut) [inline src/context.c:197]: Mh... I have to rewrute a bit, but... yes.

---

Turns out it doesn't make sense. The macro is used to iterate through arguments of some type, but it tries to extract the arguments from the value, which will result in unneccessary function calls (we really only want to count here).
I'd oppose using the macro.

Comment (matthiasbeyer) [inline src/context.c:197]: Allright!

**Comment by matthiasbeyer** [comment]
ACK.

**Pull request merged**